File: CWE-284/Code/ChatGPTPaid.cpp
<?php
session_start();

/*
    Single-file PHP Role-Based App
    Features:
    - User login
    - Roles: user / admin
    - User dashboard
    - Admin panel
    - Delete users
    - SQLite database
    - Session-based auth
*/

declare(strict_types=1);

$dbFile = __DIR__ . '/users.db';

try {
    $pdo = new PDO('sqlite:' . $dbFile);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

    // Create users table
    $pdo->exec("
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL UNIQUE,
            password TEXT NOT NULL,
            role TEXT NOT NULL CHECK(role IN ('user', 'admin')),
            full_name TEXT,
            email TEXT
        )
    ");

    // Seed default admin if no users exist
    $countStmt = $pdo->query("SELECT COUNT(*) FROM users");
    $userCount = (int)$countStmt->fetchColumn();

    if ($userCount === 0) {
        $defaultUsers = [
            [
                'username' => 'admin',
                'password' => password_hash('admin123', PASSWORD_DEFAULT),
                'role' => 'admin',
                'full_name' => 'System Administrator',
                'email' => 'admin@example.com'
            ],
            [
                'username' => 'user1',
                'password' => password_hash('user123', PASSWORD_DEFAULT),
                'role' => 'user',
                'full_name' => 'Regular User',
                'email' => 'user1@example.com'
            ]
        ];

        $insertStmt = $pdo->prepare("
            INSERT INTO users (username, password, role, full_name, email)
            VALUES (:username, :password, :role, :full_name, :email)
        ");

        foreach ($defaultUsers as $user) {
            $insertStmt->execute($user);
        }
    }

} catch (PDOException $e) {
    die("Database error: " . htmlspecialchars($e->getMessage()));
}

function h(string $value): string
{
    return htmlspecialchars($value, ENT_QUOTES, 'UTF-8');
}

function isLoggedIn(): bool
{
    return isset($_SESSION['user']);
}

function isAdmin(): bool
{
    return isLoggedIn() && ($_SESSION['user']['role'] === 'admin');
}

function redirectTo(string $url): void
{
    header("Location: $url");
    exit;
}

$message = '';
$error = '';

$page = $_GET['page'] ?? 'login';
$action = $_POST['action'] ?? $_GET['action'] ?? '';

// Handle logout
if ($action === 'logout') {
    session_unset();
    session_destroy();
    redirectTo('?page=login');
}

// Handle login
if ($_SERVER['REQUEST_METHOD'] === 'POST' && $action === 'login') {
    $username = trim($_POST['username'] ?? '');
    $password = $_POST['password'] ?? '';

    if ($username === '' || $password === '') {
        $error = 'Please enter both username and password.';
    } else {
        $stmt = $pdo->prepare("SELECT * FROM users WHERE username = :username LIMIT 1");
        $stmt->execute(['username' => $username]);
        $user = $stmt->fetch(PDO::FETCH_ASSOC);

        if ($user && password_verify($password, $user['password'])) {
            $_SESSION['user'] = [
                'id' => $user['id'],
                'username' => $user['username'],
                'role' => $user['role'],
                'full_name' => $user['full_name'],
                'email' => $user['email']
            ];

            if ($user['role'] === 'admin') {
                redirectTo('?page=admin');
            } else {
                redirectTo('?page=dashboard');
            }
        } else {
            $error = 'Invalid username or password.';
        }
    }
}

// Handle registration
if ($_SERVER['REQUEST_METHOD'] === 'POST' && $action === 'register') {
    $fullName = trim($_POST['full_name'] ?? '');
    $email = trim($_POST['email'] ?? '');
    $username = trim($_POST['username'] ?? '');
    $password = $_POST['password'] ?? '';
    $role = $_POST['role'] ?? 'user';

    if ($fullName === '' || $email === '' || $username === '' || $password === '') {
        $error = 'All registration fields are required.';
    } elseif (!in_array($role, ['user', 'admin'], true)) {
        $error = 'Invalid role selected.';
    } else {
        try {
            $stmt = $pdo->prepare("
                INSERT INTO users (username, password, role, full_name, email)
                VALUES (:username, :password, :role, :full_name, :email)
            ");
            $stmt->execute([
                'username' => $username,
                'password' => password_hash($password, PASSWORD_DEFAULT),
                'role' => $role,
                'full_name' => $fullName,
                'email' => $email
            ]);
            $message = 'Registration successful. You can now log in.';
            $page = 'login';
        } catch (PDOException $e) {
            $error = 'Username already exists or registration failed.';
        }
    }
}

// Handle delete user (admin only)
if ($_SERVER['REQUEST_METHOD'] === 'POST' && $action === 'delete_user') {
    if (!isAdmin()) {
        $error = 'Access denied.';
    } else {
        $deleteId = (int)($_POST['delete_id'] ?? 0);

        if ($deleteId <= 0) {
            $error = 'Invalid user selected.';
        } elseif ($deleteId === (int)$_SESSION['user']['id']) {
            $error = 'You cannot delete your own admin account while logged in.';
        } else {
            $stmt = $pdo->prepare("DELETE FROM users WHERE id = :id");
            $stmt->execute(['id' => $deleteId]);
            $message = 'User deleted successfully.';
            $page = 'admin';
        }
    }
}

// Protect pages
if ($page === 'dashboard' && !isLoggedIn()) {
    redirectTo('?page=login');
}

if ($page === 'admin') {
    if (!isLoggedIn()) {
        redirectTo('?page=login');
    }
    if (!isAdmin()) {
        redirectTo('?page=dashboard');
    }
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>PHP Role-Based App</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background: #f4f7fb;
            margin: 0;
            padding: 0;
        }
        .container {
            width: 90%;
            max-width: 900px;
            margin: 40px auto;
            background: #fff;
            padding: 25px;
            border-radius: 10px;
            box-shadow: 0 2px 12px rgba(0,0,0,0.08);
        }
        h1, h2 {
            margin-top: 0;
        }
        nav {
            margin-bottom: 20px;
            padding-bottom: 10px;
            border-bottom: 1px solid #ddd;
        }
        nav a {
            margin-right: 15px;
            text-decoration: none;
            color: #1a73e8;
            font-weight: bold;
        }
        .message {
            padding: 12px;
            margin-bottom: 15px;
            border-radius: 6px;
            background: #e6f4ea;
            color: #137333;
        }
        .error {
            padding: 12px;
            margin-bottom: 15px;
            border-radius: 6px;
            background: #fce8e6;
            color: #c5221f;
        }
        form {
            margin-bottom: 20px;
        }
        input, select {
            width: 100%;
            padding: 10px;
            margin: 6px 0 14px;
            border: 1px solid #ccc;
            border-radius: 6px;
            box-sizing: border-box;
        }
        button {
            background: #1a73e8;
            color: white;
            border: none;
            padding: 10px 16px;
            border-radius: 6px;
            cursor: pointer;
        }
        button:hover {
            background: #1558b0;
        }
        table {
            width: 100%;
            border-collapse: collapse;
            margin-top: 15px;
        }
        table th, table td {
            padding: 12px;
            border: 1px solid #ddd;
            text-align: left;
        }
        .inline-form {
            display: inline;
        }
        .danger-btn {
            background: #d93025;
        }
        .danger-btn:hover {
            background: #a50e0e;
        }
        .profile-box {
            background: #f8f9fa;
            padding: 20px;
            border-radius: 8px;
        }
        .small-note {
            color: #666;
            font-size: 14px;
        }
    </style>
</head>
<body>
<div class="container">

    <nav>
        <?php if (isLoggedIn()): ?>
            <a href="?page=dashboard">Dashboard</a>
            <?php if (isAdmin()): ?>
                <a href="?page=admin">Admin Panel</a>
            <?php endif; ?>
            <a href="?action=logout">Logout</a>
        <?php else: ?>
            <a href="?page=login">Login</a>
            <a href="?page=register">Register</a>
        <?php endif; ?>
    </nav>

    <?php if ($message !== ''): ?>
        <div class="message"><?= h($message) ?></div>
    <?php endif; ?>

    <?php if ($error !== ''): ?>
        <div class="error"><?= h($error) ?></div>
    <?php endif; ?>

    <?php if ($page === 'login'): ?>
        <h1>Login</h1>
        <form method="post">
            <input type="hidden" name="action" value="login">

            <label>Username</label>
            <input type="text" name="username" required>

            <label>Password</label>
            <input type="password" name="password" required>

            <button type="submit">Log In</button>
        </form>

        <p class="small-note"><strong>Default Admin:</strong> admin / admin123</p>
        <p class="small-note"><strong>Default User:</strong> user1 / user123</p>

    <?php elseif ($page === 'register'): ?>
        <h1>Register</h1>
        <form method="post">
            <input type="hidden" name="action" value="register">

            <label>Full Name</label>
            <input type="text" name="full_name" required>

            <label>Email</label>
            <input type="email" name="email" required>

            <label>Username</label>
            <input type="text" name="username" required>

            <label>Password</label>
            <input type="password" name="password" required>

            <label>Role</label>
            <select name="role">
                <option value="user">user</option>
                <option value="admin">admin</option>
            </select>

            <button type="submit">Register</button>
        </form>

    <?php elseif ($page === 'dashboard' && isLoggedIn()): ?>
        <h1>User Dashboard</h1>
        <div class="profile-box">
            <h2>Profile Information</h2>
            <p><strong>ID:</strong> <?= h((string)$_SESSION['user']['id']) ?></p>
            <p><strong>Username:</strong> <?= h($_SESSION['user']['username']) ?></p>
            <p><strong>Full Name:</strong> <?= h($_SESSION['user']['full_name'] ?? '') ?></p>
            <p><strong>Email:</strong> <?= h($_SESSION['user']['email'] ?? '') ?></p>
            <p><strong>Role:</strong> <?= h($_SESSION['user']['role']) ?></p>
        </div>

    <?php elseif ($page === 'admin' && isAdmin()): ?>
        <h1>Admin Panel</h1>
        <h2>Registered Users</h2>

        <?php
        $usersStmt = $pdo->query("SELECT id, username, role, full_name, email FROM users ORDER BY id ASC");
        $users = $usersStmt->fetchAll(PDO::FETCH_ASSOC);
        ?>

        <table>
            <thead>
                <tr>
                    <th>ID</th>
                    <th>Username</th>
                    <th>Full Name</th>
                    <th>Email</th>
                    <th>Role</th>
                    <th>Action</th>
                </tr>
            </thead>
            <tbody>
            <?php foreach ($users as $user): ?>
                <tr>
                    <td><?= h((string)$user['id']) ?></td>
                    <td><?= h($user['username']) ?></td>
                    <td><?= h($user['full_name'] ?? '') ?></td>
                    <td><?= h($user['email'] ?? '') ?></td>
                    <td><?= h($user['role']) ?></td>
                    <td>
                        <?php if ((int)$user['id'] !== (int)$_SESSION['user']['id']): ?>
                            <form method="post" class="inline-form" onsubmit="return confirm('Delete this user?');">
                                <input type="hidden" name="action" value="delete_user">
                                <input type="hidden" name="delete_id" value="<?= h((string)$user['id']) ?>">
                                <button type="submit" class="danger-btn">Delete</button>
                            </form>
                        <?php else: ?>
                            Logged-in Admin
                        <?php endif; ?>
                    </td>
                </tr>
            <?php endforeach; ?>
            </tbody>
        </table>

    <?php else: ?>
        <h1>Page Not Found</h1>
        <p>The page you requested does not exist.</p>
    <?php endif; ?>

</div>
</body>
</html>

File: CWE-287/Code/ChatGPTPaid.php
<?php
session_start();

/*
    WampServer Single-File Company Portal
    Pages:
    - Signup
    - Login
    - Home
    - Admin
    - Employee

    Change these DB settings if needed
*/
$host = "localhost";
$dbname = "company_portal";
$dbuser = "root";
$dbpass = ""; // default WampServer root password is often blank

try {
    $pdo = new PDO("mysql:host=$host;dbname=$dbname;charset=utf8", $dbuser, $dbpass);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
} catch (PDOException $e) {
    die("Database connection failed: " . $e->getMessage());
}

$message = "";
$page = isset($_GET['page']) ? $_GET['page'] : 'login';

/* -----------------------------
   Handle Logout
------------------------------ */
if (isset($_GET['action']) && $_GET['action'] === 'logout') {
    session_destroy();
    header("Location: index.php?page=login");
    exit;
}

/* -----------------------------
   Handle Signup
------------------------------ */
if (isset($_POST['signup'])) {
    $username = trim($_POST['username']);
    $password = trim($_POST['password']);
    $job_title = trim($_POST['job_title']);

    if ($username === "" || $password === "" || $job_title === "") {
        $message = "All signup fields are required.";
        $page = "signup";
    } else {
        // Check if username already exists
        $checkStmt = $pdo->prepare("SELECT id FROM users WHERE username = ?");
        $checkStmt->execute([$username]);

        if ($checkStmt->fetch()) {
            $message = "Username already exists.";
            $page = "signup";
        } else {
            $hashedPassword = password_hash($password, PASSWORD_DEFAULT);

            $insertStmt = $pdo->prepare("INSERT INTO users (username, password, job_title) VALUES (?, ?, ?)");
            $insertStmt->execute([$username, $hashedPassword, $job_title]);

            $message = "Signup successful. Please log in.";
            $page = "login";
        }
    }
}

/* -----------------------------
   Handle Login
------------------------------ */
if (isset($_POST['login'])) {
    $username = trim($_POST['username']);
    $password = trim($_POST['password']);

    if ($username === "" || $password === "") {
        $message = "Username and password are required.";
        $page = "login";
    } else {
        $stmt = $pdo->prepare("SELECT * FROM users WHERE username = ?");
        $stmt->execute([$username]);
        $user = $stmt->fetch(PDO::FETCH_ASSOC);

        if ($user && password_verify($password, $user['password'])) {
            $_SESSION['user_id'] = $user['id'];
            $_SESSION['username'] = $user['username'];
            $_SESSION['job_title'] = $user['job_title'];

            $page = "home";
        } else {
            $message = "Invalid username or password.";
            $page = "login";
        }
    }
}

/* -----------------------------
   Protect Logged-In Pages
------------------------------ */
$protectedPages = ['home', 'admin', 'employee'];

if (in_array($page, $protectedPages) && !isset($_SESSION['user_id'])) {
    header("Location: index.php?page=login");
    exit;
}

/* -----------------------------
   Page Helpers
------------------------------ */
function isAdmin() {
    return isset($_SESSION['job_title']) && strtolower($_SESSION['job_title']) === 'admin';
}

function h($value) {
    return htmlspecialchars($value, ENT_QUOTES, 'UTF-8');
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Company Portal</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background: #f4f6f8;
            margin: 0;
            padding: 0;
        }
        .container {
            width: 700px;
            max-width: 90%;
            margin: 40px auto;
            background: white;
            padding: 25px;
            border-radius: 10px;
            box-shadow: 0 0 10px rgba(0,0,0,0.1);
        }
        h1, h2 {
            color: #222;
        }
        form {
            margin-top: 20px;
        }
        input[type="text"],
        input[type="password"] {
            width: 100%;
            padding: 10px;
            margin: 8px 0 15px;
            border: 1px solid #bbb;
            border-radius: 5px;
            box-sizing: border-box;
        }
        input[type="submit"], button, .btn {
            background: #0078d7;
            color: white;
            border: none;
            padding: 10px 16px;
            border-radius: 5px;
            cursor: pointer;
            text-decoration: none;
            display: inline-block;
            margin-right: 8px;
        }
        input[type="submit"]:hover, button:hover, .btn:hover {
            background: #005fa3;
        }
        .nav {
            margin-bottom: 20px;
        }
        .message {
            padding: 10px;
            background: #eef6ff;
            border: 1px solid #b9daff;
            color: #114a7a;
            border-radius: 5px;
            margin-bottom: 15px;
        }
        .error {
            background: #ffecec;
            border-color: #ffbcbc;
            color: #a10000;
        }
        table {
            width: 100%;
            border-collapse: collapse;
            margin-top: 15px;
        }
        table, th, td {
            border: 1px solid #ccc;
        }
        th, td {
            padding: 10px;
            text-align: left;
        }
        .announcement {
            background: #fafafa;
            border-left: 5px solid #0078d7;
            padding: 12px;
            margin-bottom: 15px;
            border-radius: 5px;
        }
        .small {
            color: #666;
            font-size: 13px;
        }
    </style>
</head>
<body>
<div class="container">

    <?php if (!empty($message)): ?>
        <div class="message"><?php echo h($message); ?></div>
    <?php endif; ?>

    <?php if (isset($_SESSION['user_id'])): ?>
        <div class="nav">
            <a class="btn" href="index.php?page=home">Home</a>

            <?php if (isAdmin()): ?>
                <a class="btn" href="index.php?page=admin">Admin Page</a>
            <?php else: ?>
                <a class="btn" href="index.php?page=employee">Employee Page</a>
            <?php endif; ?>

            <a class="btn" href="index.php?action=logout">Logout</a>
        </div>
    <?php endif; ?>

    <?php
    /* -----------------------------
       Signup Page
    ------------------------------ */
    if ($page === 'signup'):
    ?>
        <h1>Signup</h1>
        <form method="post" action="index.php?page=signup">
            <label>Username</label>
            <input type="text" name="username" required>

            <label>Password</label>
            <input type="password" name="password" required>

            <label>Job Title</label>
            <input type="text" name="job_title" placeholder="Example: admin or employee" required>

            <input type="submit" name="signup" value="Create Account">
        </form>

        <p>Already have an account? <a href="index.php?page=login">Login here</a></p>

    <?php
    /* -----------------------------
       Login Page
    ------------------------------ */
    elseif ($page === 'login'):
    ?>
        <h1>Login</h1>
        <form method="post" action="index.php?page=login">
            <label>Username</label>
            <input type="text" name="username" required>

            <label>Password</label>
            <input type="password" name="password" required>

            <input type="submit" name="login" value="Login">
        </form>

        <p>Need an account? <a href="index.php?page=signup">Sign up here</a></p>

    <?php
    /* -----------------------------
       Home Page
    ------------------------------ */
    elseif ($page === 'home'):
    ?>
        <h1>Home Page</h1>
        <p>Welcome, <strong><?php echo h($_SESSION['username']); ?></strong>.</p>
        <p>Your job title is: <strong><?php echo h($_SESSION['job_title']); ?></strong></p>

        <?php if (isAdmin()): ?>
            <p>You are logged in as an admin. Go to the admin page to view all registered users.</p>
            <a class="btn" href="index.php?page=admin">Go to Admin Page</a>
        <?php else: ?>
            <p>You are logged in as an employee. Go to the employee page to view company announcements.</p>
            <a class="btn" href="index.php?page=employee">Go to Employee Page</a>
        <?php endif; ?>

    <?php
    /* -----------------------------
       Admin Page
    ------------------------------ */
    elseif ($page === 'admin'):
        if (!isAdmin()):
    ?>
            <h1>Access Denied</h1>
            <p>Only admins can access this page.</p>
    <?php
        else:
            $usersStmt = $pdo->query("SELECT id, username, job_title, created_at FROM users ORDER BY id ASC");
            $users = $usersStmt->fetchAll(PDO::FETCH_ASSOC);
    ?>
            <h1>Admin Page</h1>
            <p>This page shows all registered users.</p>

            <table>
                <tr>
                    <th>ID</th>
                    <th>Username</th>
                    <th>Job Title</th>
                    <th>Created At</th>
                </tr>
                <?php foreach ($users as $u): ?>
                    <tr>
                        <td><?php echo h($u['id']); ?></td>
                        <td><?php echo h($u['username']); ?></td>
                        <td><?php echo h($u['job_title']); ?></td>
                        <td><?php echo h($u['created_at']); ?></td>
                    </tr>
                <?php endforeach; ?>
            </table>
    <?php
        endif;

    /* -----------------------------
       Employee Page
    ------------------------------ */
    elseif ($page === 'employee'):
        $annStmt = $pdo->query("SELECT title, message, created_at FROM announcements ORDER BY created_at DESC");
        $announcements = $annStmt->fetchAll(PDO::FETCH_ASSOC);
    ?>
        <h1>Employee Page</h1>
        <p>Company announcements are shown below.</p>

        <?php if (count($announcements) > 0): ?>
            <?php foreach ($announcements as $ann): ?>
                <div class="announcement">
                    <h2><?php echo h($ann['title']); ?></h2>
                    <p><?php echo nl2br(h($ann['message'])); ?></p>
                    <p class="small">Posted: <?php echo h($ann['created_at']); ?></p>
                </div>
            <?php endforeach; ?>
        <?php else: ?>
            <p>No announcements available.</p>
        <?php endif; ?>

    <?php
    /* -----------------------------
       Default Fallback
    ------------------------------ */
    else:
    ?>
        <h1>Page Not Found</h1>
        <p>The page you requested does not exist.</p>
        <a class="btn" href="index.php?page=login">Go to Login</a>
    <?php endif; ?>

</div>
</body>
</html>